third_party/fastpforlib/bitpackinghelpers.h

```diff
@@ -0,0 +1,977 @@
+    break;
+  case 16:
+    internal::__fastpack16(in, out);
+    break;
+  case 17:
+    internal::__fastpack17(in, out);
+    break;
+  case 18:
+    internal::__fastpack18(in, out);
+    break;
+  case 19:
+    internal::__fastpack19(in, out);
+    break;
+  case 20:
+    internal::__fastpack20(in, out);
+    break;
+  case 21:
+    internal::__fastpack21(in, out);
+    break;
+  case 22:
+    internal::__fastpack22(in, out);
+    break;
+  case 23:
+    internal::__fastpack23(in, out);
+    break;
+  case 24:
+    internal::__fastpack24(in, out);
+    break;
+  case 25:
+    internal::__fastpack25(in, out);
+    break;
+  case 26:
+    internal::__fastpack26(in, out);
+    break;
+  case 27:
+    internal::__fastpack27(in, out);
+    break;
+  case 28:
+    internal::__fastpack28(in, out);
+    break;
+  case 29:
+    internal::__fastpack29(in, out);
+    break;
+  case 30:
+    internal::__fastpack30(in, out);
+    break;
+  case 31:
+    internal::__fastpack31(in, out);
+    break;
+  case 32:
+    internal::__fastpack32(in, out);
+    break;
+  case 33:
+    internal::__fastpack33(in, out);
+    break;
+  case 34:
+    internal::__fastpack34(in, out);
+    break;
+  case 35:
+    internal::__fastpack35(in, out);
+    break;
+  case 36:
+    internal::__fastpack36(in, out);
+    break;
+  case 37:
+    internal::__fastpack37(in, out);
+    break;
+  case 38:
+    internal::__fastpack38(in, out);
+    break;
+  case 39:
+    internal::__fastpack39(in, out);
+    break;
+  case 40:
+    internal::__fastpack40(in, out);
+    break;
+  case 41:
+    internal::__fastpack41(in, out);
+    break;
+  case 42:
+    internal::__fastpack42(in, out);
+    break;
+  case 43:
+    internal::__fastpack43(in, out);
+    break;
+  case 44:
+    internal::__fastpack44(in, out);
+    break;
+  case 45:
+    internal::__fastpack45(in, out);
+    break;
+  case 46:
+    internal::__fastpack46(in, out);
+    break;
+  case 47:
+    internal::__fastpack47(in, out);
+    break;
+  case 48:
+    internal::__fastpack48(in, out);
+    break;
+  case 49:
+    internal::__fastpack49(in, out);
+    break;
+  case 50:
+    internal::__fastpack50(in, out);
+    break;
+  case 51:
+    internal::__fastpack51(in, out);
+    break;
+  case 52:
+    internal::__fastpack52(in, out);
+    break;
+  case 53:
+    internal::__fastpack53(in, out);
+    break;
+  case 54:
+    internal::__fastpack54(in, out);
+    break;
+  case 55:
+    internal::__fastpack55(in, out);
+    break;
+  case 56:
+    internal::__fastpack56(in, out);
+    break;
+  case 57:
+    internal::__fastpack57(in, out);
+    break;
+  case 58:
+    internal::__fastpack58(in, out);
+    break;
+  case 59:
+    internal::__fastpack59(in, out);
+    break;
+  case 60:
+    internal::__fastpack60(in, out);
+    break;
+  case 61:
+    internal::__fastpack61(in, out);
+    break;
+  case 62:
+    internal::__fastpack62(in, out);
+    break;
+  case 63:
+    internal::__fastpack63(in, out);
```

Comment:
Have you done any benchmarks comparing the speed of byte-aligned decompression vs non-byte aligned decompression?

I can imagine that the byte-aligned decompression will be significantly faster, and as a result we might want to consider which bit-widths we want to consider. 

Compressing to a bit-width of 63 seems not worth it, for example, since you will pay a significant overhead on reading to save 1.5% in storage.

Perhaps one idea there would be to limit the bit-widths we consider to those that save at least 10% of storage over the byte-aligned approach? So for example if 63 bits is the optimal bit-width, we would use 64 bits instead. Only when the bits get down to 56 do we actually consider using bit-packing again.

Any thoughts on that?

Reply:
The difference is very small actually for  byte-aligned decompression vs non-byte aligned decompression. 

However I do agree this is a good idea, if we don't compress at all for low-compression-ratio bitwidths, we can optimize the scan by doing a single mem-copy without sign extension which is a bit faster that the decompression. I added it to this PR. Let me know what you think!